### ISSUE TITLE
Space-saving automatic data files compression for archiving old backups

### ISSUE THREAD

**Issue by nekohayo** [getting-things-gnome/gtg#443]
GTG has a built-in automatic daily backup system for user data.
The only thing is, after a while (years), the collection of backups starts becoming a bit heavy (even though disk space is cheap these days) and it also feels a bit messy when you look at the data folder as a user.

Personally I have found that instead of removing old backup files, I'm better off just zipping them (that way they are always around if I ever need to do some archeological digging, and it allows users to keep the data around for the eventual prophecized day where GTG could process it all into some sort of amazing local analytics thing, like #424 and/or #435). Once the filesize is minimized through compression, keeping the data around has no cost. To put things in perspective:
* my data files from November 2017 to September 2019 weighted 277 MB, and once compressed into a .tar.xz they become a single 2.2 MB file;
* similarly, my data from February 2014 to November 2017, weighted 418.4 MB on disk but shrunk down to 5.7 MB once put into a .tar.xz.

If someone wants a fun and easy project/challenge to make a plugin to offer a GUI to do that, it would be a welcome addition. There are various ways to do this:
* An action that shows up in the cleanup section of the preferences GUI that says "Your data files daily backups are currently using %d MB of disk space. You can compress and archive old backups to save approximately 90% of that space. ((Compress now))".
* The same thing, but in a GtkInfobar that gets shown in the main window (rather than the preferences GUI) everytime the data weighs more than 250 MB (maybe the threshold size could be configurable by users in the plugin's GUI settings if they somehow need something more, or less, aggressive)
* Just doing it automatically in the background when the threshold is reached, or simply packing one compressed archive per year, archiving whatever files are present from the previous (not the current) year, so compressed archives would clearly be organized yearly.

Ideally, any such processing operation should be async and not block the GTK mainloop.

Alternatively, there's the theoretical question, "what if GTG transparently used compression on individual files at all times as a core feature?" But the downsides of that would be that you'd lose the ability to use diff or Meld to compare files, and there could be potentially significant overhead with (un)compressing all the time.

**Comment by pentux-GitHub** [comment]
Good news: with the new 0.7 core this became a small, safe patch, and I have it working with tests (draft PR at #1317). Before un-drafting it, here is the design for your validation, since it bends the original request on two points.

**Core, not a plugin.** The issue suggested a plugin, which made sense in 2020 when the core was frozen. Today this is ~40 lines in the existing `write_backups` path. Backup lifecycle feels like a property of the backup system itself rather than an optional feature — and keeping it in core means every user's archive works without discovering a plugin.

**Logrotate pattern, archives kept forever.** Daily backups stay plain for 30 days (trivial to inspect and restore by hand — a virtue in a data-recovery situation), then get gzipped in place. XML compresses to a few percent of its size, so `*.bak.gz` archives are never removed automatically: that is the "space-saving archiving" this issue asked for, history preserved rather than pruned.

**Safety rules**, since this code is the data safety net: the compressed copy is read back and byte-compared against the original before the plain file is removed; on any failure the plain file stays and archival is retried next run. The automatic restore path (`find_and_load_file`) is untouched — it only relies on the plain rotating `.bak.0-N` copies; restoring from an archive is a manual `gunzip` away.

One heads-up: while writing this I found that the current age-based purge targets the wrong directory entirely (#1316), so the fix for that comes first in the same draft PR.

Happy to adjust any of the above — the draft is there to be shaped.

**Comment by nekohayo** [comment]
Yes, I'd be happy if this is a built-in core feature, actually! In fact, recently I was wondering why GTG's format isn't always compressed anyway, even individual files, since it compresses so well and the cost is probably negligible.

Through https://gitlab.gnome.org/GNOME/nautilus/-/work_items/1936 I have recently discovered the existence of [zstandard](https://en.wikipedia.org/wiki/Zstd) and, according to my observations, it lets you compress either as fast as (or faster than) gzip while providing smaller filesizes, or can compress harder between the spectrum of gzip and xz depending on what middle-ground setting you use. Apparently zstd is quite widespread on Linux distros (and if isn't, maybe we could bundle it), so would it be better to standardize on zstd rather than gzip while we're at it? See also https://github.com/virtualmin/virtualmin-gpl/issues/1274 for a similar argument I made for Virtualmin to simplify the choices/dependencies.
